src/Microsoft.Diagnostics.Monitoring.StartupHook/Exceptions/Identification/ExceptionIdentifierCache.cs

```diff
@@ -121,6 +135,57 @@ public ulong GetOrAdd(Module module)
             return moduleId;
         }
 
+        /// <summary>
+        /// Gets the identifier of the <see cref="StackFrame"/> instance and
+        /// caches its data if it has not been encountered yet.
+        /// </summary>
+        /// <returns>
+        /// An identifier that uniquely identifies the <see cref="StackFrame"/> in this cache.
+        /// </returns>
+        public ulong GetOrAdd(StackFrame frame)
+        {
+            StackFrameIdentifier identifier = new(
+                AddOrDefault(frame.GetMethod()),
+                frame.GetILOffset());
+
+            if (_stackFrameIds.TryGetValue(identifier, out ulong frameId))
+                return frameId;
+
+            frameId = Interlocked.Increment(ref _nextStackFrameId);
+
+            ulong actualFrameId = _stackFrameIds.GetOrAdd(identifier, frameId);
+
+            if (actualFrameId == frameId)
```

Comment:
Can you add some comments around this scenario? I'm assuming this is to handle concurrency.

Reply:
I ended up refactoring all of the usages of this pattern into one method that gets called by each concept. The comments have been added to that method to explain in great detail how it's handling concurrent inserts into the dictionary.